data-structures/HeaderbasedCircularDoublyLinkedList.cpp

```diff
@@ -0,0 +1,216 @@
+    if(head->next == NULL) {
+        cout<<"List is already empty, can't perform insertion."<<endl;
+        return head;
+    }
+    HCDNODE *q=head->next;
+    while(q->info!=key) {
+        q = q->next;
+        if(q == head->next) {
+            cout<<"key "<<key<<" not found, can't perform insertion."<<endl;
+            return head;
+        }
+    }
+    HCDNODE *p = (HCDNODE *)malloc(sizeof(HCDNODE));
+    p->info = val;
+    p->prev = q;
+    p->next = q->next;
+    p->next->prev = p;
+    q->next = p;
+    head->info++;
+    return head;
+}
+
+HCDNODE *insertBefore(HCDNODE *head, int val, int key) {
+    if(head->next == NULL) {
+        cout<<"List is already empty, can't perform insertion."<<endl;
+        return head;
+    }
+    HCDNODE *q=head->next;
+    while(q->info!=key) {
+        q = q->next;
+        if(q == head->next) {
+            cout<<"key "<<key<<" not found, can't perform insertion."<<endl;
+            return head;
+        }
+    }
+    HCDNODE *p = (HCDNODE *)malloc(sizeof(HCDNODE));
+    p->info = val;
+    p->prev = q->prev;
+    p->next = q;
+    p->prev->next = p;
+    q->prev = p;
+    if(q == head->next)
+        head->next = p;
+    head->info++;
+    return head;
+}
+
+HCDNODE *deletion(HCDNODE *head, int key, bool isFirst=false) {
+    if(head->next == NULL) {
+        cout<<"List is already empty."<<endl;
+        return head;
+    }
+    if(key == -1) { //means we want to delete either first or last element of the list
+        if(head->next->next == head->next) {
+            free(head->next);
+            head->next = NULL;
+            head->info--;
+            return head;
+        }
+        if(isFirst) {
+            HCDNODE *p = head->next;
+            head->next->prev->next = head->next->next;
+            head->next->next->prev = head->next->prev;
+            head->next = head->next->next;
+            free(p);
+            head->info--;
+            return head;
+        } else {
+            HCDNODE *q = head->next->prev;
+            q->prev->next = q->next;
+            q->next->prev = q->prev;
+            free(q);
+            head->info--;
+            return head;
+        }
+    }
+    HCDNODE *q=head->next;
+    while(q->info!=key) {
+        q = q->next;
+        if(q == head->next) {
+            cout<<"Element not found, can't perform deletion."<<endl;
+            return head;
+        }
+    }
+    if(q->next == q) {
+        free(q);
+        head->next = NULL;
+        head->info--;
+        return head;
+    }
+    q->prev->next = q->next;
+    q->next->prev = q->prev;
+    if(q==head->next)
+        head = q->next;
+    free(q);
+    head->info--;
+    return head;
+}
+
+void display(HCDNODE *head) {
+    if(head->next == NULL) {
+        cout<<"List is empty."<<endl;
+        return ;
+    }
+    cout<<"Header = "<<head->info<<", List = ";
+    HCDNODE *q = head->next;
+    do {
+        cout<<q->info<<" ";
+        q = q->next;
+    }while(q!=head->next);
```

Comment:
Please use this conventions https://google.github.io/styleguide/cppguide.html#Loops_and_Switch_Statements